### PR TITLE
HV-1566 Cache type variable index for performance reasons

### PR DESCRIPTION
**Pull request by marko-bekhta**
- https://hibernate.atlassian.net/browse/HV-1566

and same results as in the JIRA ticket:
```bash
current master:

# Run complete. Total time: 00:01:52

Benchmark                                                                                   Mode  Cnt    Score    Error   Units
CascadedValidation.testCascadedValidation                                                  thrpt   20  160.152 ± 11.923  ops/ms
CascadedWithLotsOfItemsAndMoreConstraintsValidation.testCascadedValidationWithLotsOfItems  thrpt   20  471.770 ±  8.797   ops/s
CascadedWithLotsOfItemsValidation.testCascadedValidationWithLotsOfItems                    thrpt   20  633.382 ± 11.182   ops/s

with changes

# Run complete. Total time: 00:01:50

Benchmark                                                                                   Mode  Cnt    Score    Error   Units
CascadedValidation.testCascadedValidation                                                  thrpt   20  170.439 ± 15.699  ops/ms
CascadedWithLotsOfItemsAndMoreConstraintsValidation.testCascadedValidationWithLotsOfItems  thrpt   20  471.262 ± 20.874   ops/s
CascadedWithLotsOfItemsValidation.testCascadedValidationWithLotsOfItems                    thrpt   20  703.661 ± 39.131   ops/s
```

### CODE REVIEW

File: engine/src/main/java/org/hibernate/validator/internal/engine/ValueContext.java [hunk 1]
```diff
@@ -171,20 +171,19 @@ public final void markCurrentPropertyAsIterableAndSetIndex(Integer index) {
 	 * Sets the container element information.
 	 *
 	 * @param containerClass the class of the container
-	 * @param typeParameter the actual type parameter (i.e. not one of our internal type parameters)
+	 * @param typeParameterIndex the index of actual type parameter (i.e. not one of our internal type parameters)
```
Comment (gsmet) [inline engine/src/main/java/org/hibernate/validator/internal/engine/ValueContext.java]: the index of **the** actual type parameter

**Comment by gsmet** [comment]
OK, this looks good and it's a good thing to improve the container element case. I don't think we have benchmarked it a lot. Will merge.

File: engine/src/main/java/org/hibernate/validator/internal/engine/ValueContext.java [hunk 2]
```diff
@@ -171,20 +171,19 @@ public final void markCurrentPropertyAsIterableAndSetIndex(Integer index) {
 	 * Sets the container element information.
 	 *
 	 * @param containerClass the class of the container
-	 * @param typeParameter the actual type parameter (i.e. not one of our internal type parameters)
+	 * @param typeParameterIndex the index of the actual type parameter (i.e. not one of our internal type parameters)
```
Comment (gunnarmorling) [inline engine/src/main/java/org/hibernate/validator/internal/engine/ValueContext.java]: Seems the (...) part is superfluous now?

Comment (marko-bekhta) [reply engine/src/main/java/org/hibernate/validator/internal/engine/ValueContext.java]: yes it seems so, as we just pass an already calculated index to this method. Removed the (...) part.

**Comment by gsmet** [comment]
Merged with a few further minor tweaks: in our getter, we usually don't use `this.` (at least not in the modified classes) so I removed them.

Thanks!

**Comment by marko-bekhta** [comment]
> we usually don't use this.

Ahh, ok. That's just how the IDE generates them. Will fix that for HV settings :)

> wondering if we shouldn't introduce our own structure (TypeParameter?) wrapping the values

That's actually very close to what I have in mind for a different PR I wanted to discuss. I was just waiting while we finish with the opened ones :)

**Pull request closed without merging**